crates/sui-core/src/epoch/reconfiguration.rs

```diff
@@ -225,4 +256,109 @@ where
         self.net.store(new_net);
         Ok(())
     }
+
+    async fn wait_for_epoch_cert(
+        &self,
+        epoch_id: EpochId,
+        old_committee: &Committee,
+        timeout_until_quorum: Duration,
+    ) -> SuiResult<CertifiedEpoch> {
+        #[derive(Default)]
+        struct Signatures {
+            total_stake: StakeUnit,
+            sigs: Vec<(AuthorityName, AuthoritySignature)>,
+        }
+        #[derive(Default)]
+        struct Summaries {
+            bad_weight: StakeUnit,
+            signed: BTreeMap<EpochInfoDigest, Signatures>,
+            cert: Option<CertifiedEpoch>,
+            errors: Vec<(AuthorityName, SuiError)>,
+        }
+        let initial_state = Summaries::default();
+        let net = self.net.load();
+        let threshold = net.committee.quorum_threshold();
+        let validity = net.committee.validity_threshold();
+        let final_state = net
+            .quorum_map_then_reduce_with_timeout(
+                initial_state,
+                |_name, client| {
+                    Box::pin(async move {
+                        client
+                            .handle_epoch(EpochRequest { epoch_id: Some(epoch_id) })
```

Comment:
So here, do you want to request an epoch_id or do you want a variant to produce the "latest" one. 

Basically this function is all or nothing: it requires to have a quorum at epoch_id, or it will fail. Further it may fail because of two many error, or it may fail because not enough authorities are at this epoch yet. So this is confusing.

Reply:
Not sure if I understand, what is the confusing part? At this point we must wait for a quorum of validators to have epoch info at epoch_id.